`cs fetch` caches a POM with no artifact, and fails to download on subsequent runs with different repositories

The maven package `javax.media:jai_core:1.1.3` [exists in an incomplete state](https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/) on `repo1.maven.org/maven2`, which is one of Coursier's default repositories.

To satisfy dependencies from `org.osgeo`, the package is hosted at `https://repo.osgeo.org/repository/release`. The JAR is available from there.

We've discovered that you can get Coursier into a state where it will permanently fail to download this artifact.

## Symptoms

There are three separate problematic behaviours that Coursier is exhibiting here:

1. If `repo1.maven.org` comes before `repo.osgeo.org` in Coursier's resolve order, Coursier will download the POM at `repo1.maven.org`, rather than trying the other configured repository to see if it possible to resolve there
2. If `--no-default` is not specified, `repo1.maven.org` is _always_ tested before user-supplied repositories, which leads to unexpected behaviour for an inexperienced user (re-ordering the repositories list will not do anything until `--no-default` is supplied, which will break other resolves)
3. Coursier _eagerly caches_ the first POM it finds for a given coordinate, regardless of whether a JAR can be fetched next to that POM, and regardless of which repositories are configured on subsequent runs. Consequently, once `javax.media:jai_core:1.1.3` is attempted to be fetched from `repo1.maven.org`, subsequent Coursier runs will always fail to fetch the JAR, regardless of what repositories are fetched. The current work-around for this is to blow away the entire Coursier cache, which is undesirable


## Expected behaviours

1. Coursier should try all configured repositories until an artifact is downloaded. Downloading just the POM for that artifact is not sufficient to show success.
2. Coursier should check user-supplied repositories before the default repositories
3. Coursier should not cache this POM because it does not supply a usable artifact, and causes Coursier to actively get in the way of fixing these symptoms
4. (BONUS) Coursier should cache each artifact according to the repository it was found at, so that if the repositories list ever changes (to get a different packaging of an artifact maybe?) it will download it more eagerly.


## How to reproduce


1. Work in a tmpdir: `mkdir /tmp/coursier && cd /tmp/coursier`

2.
Delete the coursier download cache, to start with clean state. E.g., on MacOS: `rm -rf ~/Library/Caches/Coursier/v1/`.

5. 
Fetch the broken artifact from the bad repository:
```
$ cs fetch -r=https://repo1.maven.org/maven2 --no-default --json-output-file=coursier_report.json javax.media:jai_core:1.1.3

$ cat coursier_report.json | jq -M .dependencies[0].file
null
```

6.
Try and refetch with the good repository:
```
$ rm -f coursier_report.json  # Just to be sure that isn't confusing us somehow
$ cs fetch -r=https://repo.osgeo.org/repository/release -r=https://repo1.maven.org/maven2 --no-default --json-output-file=coursier_report.json javax.media:jai_core:1.1.3

$ cat coursier_report.json | jq -M .dependencies[0].file
null
```

7.
Delete the coursier cache again, as above.

8.
Refetch with the good repository, as above:
```
$ rm -f coursier_report.json 
$ cs fetch -r=https://repo.osgeo.org/repository/release -r=https://repo1.maven.org/maven2 --no-default --json-output-file=coursier_report.json javax.media:jai_core:1.1.3

$ cat coursier_report.json | jq -M .dependencies[0].file
"/Users/benjyw/Library/Caches/Coursier/v1/https/repo.osgeo.org/repository/release/javax/media/jai_core/1.1.3/jai_core-1.1.3.jar"
```

I've been troubleshooting a SBT build failure and this seems like a candidate for the root cause.

In our scenario, access to Maven central is mediated by a proxy that gives 403 errors for artifacts that are yet to pass automated scanning for licence compliance etc. It seems that this process does not atomically process dependencies, and a normal transitional state serves up the .pom (with packaging=jar) but denies access to the JARs.

This effectively looks like the incomplete javax.media:jai_core:1.1.3 artifact you pointed out.

SBT uses Coursier to fetch dependencies of the build itself, and, later, of the user code. We notices that in the first phase this incomplete does not result in a resolution error, and a later error ensues when the classpath for the build definition is missing a required JAR.

AFAICT, Coursier should report an error when `packaging=jar` but the JAR is not available.

Analagous resolve Maven:

```xml
    <project>
      <modelVersion>4.0.0</modelVersion>
     
      <groupId>com.mycompany.app</groupId>
      <artifactId>my-app</artifactId>
      <version>1</version>
      <dependencies>
        <dependency>
          <groupId>java.media</groupId>
          <artifactId>jai_core</artifactId>
          <version>1.1.3</version>
        </dependency>
      </dependencies>  
    </project>
```

```
 mvn install
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=300M; support was removed in 8.0
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------------< com.mycompany.app:my-app >----------------------
[INFO] Building my-app 1
[INFO] --------------------------------[ jar ]---------------------------------
Downloading from central: https://repo.maven.apache.org/maven2/java/media/jai_core/1.1.3/jai_core-1.1.3.pom
[WARNING] The POM for java.media:jai_core:jar:1.1.3 is missing, no dependency information available
Downloading from central: https://repo.maven.apache.org/maven2/java/media/jai_core/1.1.3/jai_core-1.1.3.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.411 s
[INFO] Finished at: 2022-11-10T08:08:53+10:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project my-app: Could not resolve dependencies for project com.mycompany.app:my-app:jar:1: Could not find artifact java.media:jai_core:jar:1.1.3 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

```
$ rm -rf  ~/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/javax/media/jai_core
$ cs fetch -v -r=https://repo1.maven.org/maven2 --no-default  javax.media:jai_core:1.1.3; echo $?) | tee out.txt

  Dependencies:
javax.media:jai_core:1.1.3:
Downloading https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.pom
Downloaded https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.pom
Downloading https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.jar
Downloaded https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.jar
Downloading https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.jar.sha1
Downloaded https://repo1.maven.org/maven2/javax/media/jai_core/1.1.3/jai_core-1.1.3.jar.sha1
0
```

--- a/.gitmodules
+++ b/.gitmodules
@@ -1,7 +1,6 @@
 [submodule "tests/metadata"]
 	path = modules/tests/metadata
 	url = https://github.com/coursier/test-metadata.git
-	branch = failure-isnt-optional
 [submodule "modules/directories"]
 	path = modules/directories
 	url = https://github.com/coursier/directories-jvm.git

--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,7 @@
 [submodule "tests/metadata"]
 	path = modules/tests/metadata
 	url = https://github.com/coursier/test-metadata.git
+	branch = failure-isnt-optional
 [submodule "modules/directories"]
 	path = modules/directories
 	url = https://github.com/coursier/directories-jvm.git

--- a/modules/core/shared/src/main/scala/coursier/maven/MavenRepositoryInternal.scala
+++ b/modules/core/shared/src/main/scala/coursier/maven/MavenRepositoryInternal.scala
@@ -422,7 +422,7 @@ private[coursier] class MavenRepositoryInternal(
           ) -> optional
       }
 
-      val allPubs = packagingPublicationOpt.map(_ -> true).toSeq ++ extraPubs
+      val allPubs = packagingPublicationOpt.map(_ -> false).toSeq ++ extraPubs
       val optional = allPubs
         .groupBy(_._1)
         .view

--- a/modules/coursier/jvm/src/test/scala/coursier/ArtifactsTests.scala
+++ b/modules/coursier/jvm/src/test/scala/coursier/ArtifactsTests.scala
@@ -322,13 +322,18 @@ object ArtifactsTests extends TestSuite {
           .future()
       }
 
-      val artifacts = Artifacts.artifacts(res, Set.empty, None, None, true).map(_._3).distinct
-      val groupedArtifacts = Artifacts.groupArtifacts(artifacts)
+      val databindUrl =
+        "https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.6.7.2/jackson-databind-2.6.7.2.jar"
+
+      val artifacts   = Artifacts.artifacts(res, Set.empty, None, None, true).map(_._3).distinct
+      val databindOpt = artifacts.find(_.url == databindUrl)
+      assert(databindOpt.isDefined)
+      val groupedArtifacts = Artifacts.groupArtifacts(artifacts :+ databindOpt.get)
 
       assert(groupedArtifacts.length == 2)
 
       val expectedDuplicatedUrls = Set(
-        "https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.6.7.2/jackson-databind-2.6.7.2.jar"
+        databindUrl
       )
 
       val firstGroupUrls = groupedArtifacts.head.map(_.url).toSet

--- a/modules/coursier/jvm/src/test/scala/coursier/FetchTests.scala
+++ b/modules/coursier/jvm/src/test/scala/coursier/FetchTests.scala
@@ -5,6 +5,7 @@ import java.io.File
 import coursier.core.{Activation, Configuration, Extension}
 import coursier.params.ResolutionParams
 import coursier.ivy.IvyRepository
+import coursier.util.Task
 import utest._
 
 import scala.async.Async.{async, await}
@@ -333,6 +334,33 @@ object FetchTests extends TestSuite {
       }
     }
 
-  }
+    test("jai_core") {
+      test("ko") {
+        val cache1 = cache match {
+          case cache: coursier.cache.MockCache[Task] =>
+            cache.withDummyArtifact(_ => false)
+        }
+        try {
+          val res = fetch
+            .withCache(cache1)
+            .addDependencies(dep"javax.media:jai_core:1.1.3")
+            .run()
+          println(s"res=$res")
+          assert(false)
+        }
+        catch { case _: coursier.error.FetchError.DownloadingArtifacts => () }
+      }
 
+      test("ok") - async {
+        val res = await {
+          val osgeo = MavenRepository("https://repo.osgeo.org/repository/release")
+          fetch
+            .withRepositories(Seq(osgeo, Repositories.central))
+            .addDependencies(dep"javax.media:jai_core:1.1.3")
+            .future()
+        }
+        assert(res(0) != null)
+      }
+    }
+  }
 }

